### PR TITLE
ColdDB state_root test

### PR DESCRIPTION
**Pull request by posvyatokum**
Creating a test to check that State column in cold storage is consistent.
It traverses Trie and makes sure that every node is present in cold db. 
We can pass max-depth or max-count to the tool to make sure that test is running reasonable time.
Test is randomised in a sense that children are traversed in the random order, so if max-count is passed, then the set of checked nodes may vary from run to run.

Because this test is intended to run on an actual node as a background check, introducing creation of checkpoints to public interface of the opener (open to literally any other idea here). Node still has to be stopped for creation of checkpoints, but it is a quick process.

Tool can be used to check Trie from a given state_root, but intended use is to provide some random (or not random) height.
Then previous state root of all chunks of that block will be checked (it is just easy to get to them code-wise, no need for shard ids of any kind).

### CODE REVIEW

File: tools/cold-store/src/cli.rs [hunk 1]
```diff
@@ -384,3 +424,180 @@ impl PrepareHotCmd {
...
+                    prune_state.down();
+                    // Visit a child node
+                    Self::check_trie(store, &child, prune_state, prune_condition)?;
+                    // Record in prune state that we are returning from a child node
+                    prune_state.up();
+                }
+            },
+            Err(_) => {
+                tracing::debug!(target: "check_trie", "Reached leaf node");
+                return Ok(());
```
Comment (wacban) [inline tools/cold-store/src/cli.rs:589]: This seems wrong, should you return an error from here?

File: tools/cold-store/src/cli.rs [hunk 2]
```diff
@@ -384,3 +424,180 @@ impl PrepareHotCmd {
...
+            }
+        }
+        Ok(())
+    }
+
+    fn read_state<'a>(
+        store: &'a Store,
+        key: &'a [u8],
```
Comment (wacban) [inline tools/cold-store/src/cli.rs:597]: nit: trie_key

**Review by wacban — CHANGES_REQUESTED** [review]
looks good overall, mostly nits but please clarify the error handling in check_trie

File: core/store/src/db/rocksdb/snapshot.rs [hunk 3]
```diff
@@ -18,11 +18,11 @@ use crate::Temperature;
 /// informational messages pointing where the snapshot resides and how to
 /// recover data from it.
 #[derive(Debug)]
-pub(crate) struct Snapshot(Option<std::path::PathBuf>);
+pub struct Snapshot(pub Option<std::path::PathBuf>);
```
[the remark this answers is not in the document]
Comment (posvyatokum) [reply core/store/src/db/rocksdb/snapshot.rs:21]: https://doc.rust-lang.org/book/ch05-01-defining-structs.html#using-tuple-structs-without-named-fields-to-create-different-types

File: tools/cold-store/src/cli.rs [hunk 4]
```diff
@@ -384,3 +424,180 @@ impl PrepareHotCmd {
...
+                    .map(|chunk| {
+                        cold_store
+                            .get_ser::<near_primitives::sharding::ShardChunk>(
+                                DBCol::Chunks,
+                                chunk.chunk_hash().as_bytes(),
+                            )
+                            .unwrap()
+                            .unwrap()
+                            .take_header()
+                            .prev_state_root()
```
[the remark this answers is not in the document]
Comment (posvyatokum) [reply tools/cold-store/src/cli.rs]: Just a quickest way to get to any relevant state_root. To get to state roots of chunks themselves you need shard_uids and it is just an extra step that doesn't really change the test itself.

File: tools/cold-store/src/cli.rs [hunk 5]
```diff
@@ -384,3 +441,196 @@ impl PrepareHotCmd {
...
+    pub fn up(&mut self) {
+        self.depth -= 1;
+    }
+}
+
+#[derive(clap::Args)]
+struct CheckStateRootCmd {
+    /// Maximum depth (measured in number of nodes, not trie key length) for checking trie.
+    #[clap(long)]
+    max_depth: Option<u8>,
```
Comment (wacban) [inline tools/cold-store/src/cli.rs:555]: Just out of curiosity, why as small as u8?

**Review by wacban — APPROVED** [review]
lgtm

**Pull request merged**